File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
    stack<long long> st;
    long long minEle = 0;
public:
    MinStack() {
        
    }
    
    void push(int val) {
        if(st.empty()){
            st.push(val);
            minEle = val;
        }
        else{
            if(val >= minEle)
                st.push(val);
            else{
                st.push(2ll * val - minEle);
                minEle = val;
            }
        }
    }
    
    void pop() {
        if(st.empty())
            return;
        long long val = st.top();
        if(val < minEle){
            minEle = 2ll * minEle - val;
        }
        st.pop();
    }
    
    int top() {
        if(st.empty())
            return -1;
        else{
            long long val = st.top();
            if(val >= minEle)
                return val;
            else{
                return minEle;
            }
        }
    }
    
    int getMin() {
        if(st.empty())
            return -1;
        return minEle;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Deque<Long> stack;
    long minVal;

    public MinStack() {
        stack = new ArrayDeque<>();
    }
    
    public void push(int val) {
        long value = val;

        if (stack.isEmpty()) {
            stack.push(value);
            minVal = value;
        } else if (value >= minVal) {
            stack.push(value);
        } else {
            stack.push(2 * value - minVal);
            minVal = value;
        }
    }
    
    public void pop() {
        long val = stack.pop();
        if (val < minVal) {
            minVal = 2 * minVal - val;
        }
    }
    
    public int top() {
        long val = stack.peek();
        if (val < minVal)
            return (int) minVal;
        return (int) val;
    }
    
    public int getMin() {
        return (int) minVal;
    }
}
